sqld/src/rpc/replication_log.rs

```diff
@@ -31,52 +34,30 @@ impl ReplicationLogService {
             replicas_with_hello: RwLock::new(HashSet::<SocketAddr>::new()),
         }
     }
+}
 
-    fn stream_pages(&self, start_id: FrameNo) -> ReceiverStream<Result<Frame, Status>> {
-        let logger = self.logger.clone();
-        let (sender, receiver) = tokio::sync::mpsc::channel(64);
-        tokio::task::spawn_blocking(move || {
-            let mut offset = start_id;
-            loop {
-                // FIXME: add buffering of the log frames.
-                let log_file = logger.log_file.read();
-                match log_file.frame_bytes(offset) {
-                    Err(LogReadError::Ahead) => break,
-                    Ok(data) => {
-                        // release lock asap
-                        drop(log_file);
-                        if let Err(e) = sender.blocking_send(Ok(Frame { data })) {
-                            tracing::error!("failed to send frame: {e}");
-                            break;
-                        }
-                        offset += 1;
-                    }
-                    Err(LogReadError::SnapshotRequired) => {
-                        let _ = sender.blocking_send(Err(Status::new(
-                            tonic::Code::FailedPrecondition,
-                            NEED_SNAPSHOT_ERROR_MSG,
-                        )));
-                        break;
-                    }
-                    Err(LogReadError::Error(e)) => {
-                        if let Err(e) = sender
-                            .blocking_send(Err(Status::new(tonic::Code::Internal, e.to_string())))
-                        {
-                            tracing::error!("failed to send frame: {e}");
-                        }
-                        break;
-                    }
-                }
-            }
-        });
+type ConvertStreamOutputFn = fn(Result<Bytes, LogReadError>) -> Result<Frame, Status>;
+type RpcFrameStream = futures::stream::Map<FrameStream, ConvertStreamOutputFn>;
 
-        ReceiverStream::new(receiver)
+fn map_frame_stream_output(r: Result<Bytes, LogReadError>) -> Result<Frame, Status> {
+    match r {
+        Ok(data) => Ok(Frame { data }),
+        Err(LogReadError::SnapshotRequired) => Err(Status::new(
+            tonic::Code::FailedPrecondition,
+            NEED_SNAPSHOT_ERROR_MSG,
+        )),
+        Err(LogReadError::Error(e)) => Err(Status::new(tonic::Code::Internal, e.to_string())),
+        // this error should be caught before, but we handle it nicely anyways
+        Err(LogReadError::Ahead) => Err(Status::new(
+            tonic::Code::OutOfRange,
+            "frame not yet available",
+        )),
     }
 }
 
 #[tonic::async_trait]
 impl ReplicationLog for ReplicationLogService {
-    type LogEntriesStream = ReceiverStream<Result<Frame, Status>>;
+    type LogEntriesStream = RpcFrameStream;
     type SnapshotStream = ReceiverStream<Result<Frame, Status>>;
```

Comment:
Nit: would it make sense to use boxed streams (probably `futures::stream::BoxStream`) instead of explicitly spelling out the static types of returned streams?